tendermint/src/proposal.rs

```diff
@@ -0,0 +1,261 @@
+
+mod canonical_proposal;
+mod msg_type;
+mod sign_proposal;
+
+pub use self::canonical_proposal::CanonicalProposal;
+pub use msg_type::Type;
+pub use sign_proposal::{SignProposalRequest, SignedProposalResponse};
+
+use crate::block::{Height, Id as BlockId, Round};
+use crate::chain::Id as ChainId;
+use crate::consensus::State;
+use crate::Signature;
+use crate::Time;
+use crate::{Error, Kind};
+use bytes::BufMut;
+use std::convert::{TryFrom, TryInto};
+use tendermint_proto::types::Proposal as RawProposal;
+use tendermint_proto::{DomainType, Error as DomainTypeError};
+
+/// Proposal
+#[derive(Clone, PartialEq, Debug)]
+pub struct Proposal {
+    /// Proposal message type
+    pub msg_type: Type,
+    /// Height
+    pub height: Height,
+    /// Round
+    pub round: Round,
+    /// POL Round
+    pub pol_round: Option<Round>,
+    /// Block ID
+    pub block_id: Option<BlockId>,
+    /// Timestamp
+    pub timestamp: Option<Time>,
+    /// Signature
+    pub signature: Signature,
+}
+
+impl DomainType<RawProposal> for Proposal {}
+
+impl TryFrom<RawProposal> for Proposal {
+    type Error = Error;
+
+    fn try_from(value: RawProposal) -> Result<Self, Self::Error> {
+        if value.pol_round < -1 {
+            return Err(Kind::NegativePolRound.into());
+        }
+        let pol_round = match value.pol_round {
+            -1 => None,
+            n => Some(Round::try_from(n)?),
+        };
+        Ok(Proposal {
+            msg_type: value.r#type.try_into()?,
+            height: value.height.try_into()?,
+            round: value.round.try_into()?,
+            pol_round,
+            block_id: value.block_id.map(TryInto::try_into).transpose()?,
+            timestamp: value.timestamp.map(TryInto::try_into).transpose()?,
+            signature: value.signature.try_into()?,
+        })
+    }
+}
+
+impl From<Proposal> for RawProposal {
+    fn from(value: Proposal) -> Self {
+        RawProposal {
+            r#type: value.msg_type.into(),
+            height: value.height.into(),
+            round: value.round.into(),
+            pol_round: value.pol_round.map_or(-1, Into::into),
+            block_id: value.block_id.map(Into::into),
+            timestamp: value.timestamp.map(Into::into),
+            signature: value.signature.into(),
+        }
+    }
+}
+
+impl Proposal {
+    /// Create signable bytes from Proposal.
+    pub fn to_signable_bytes<B>(
+        &self,
+        chain_id: ChainId,
+        sign_bytes: &mut B,
+    ) -> Result<bool, DomainTypeError>
+    where
+        B: BufMut,
+    {
+        CanonicalProposal::new(self.clone(), chain_id).encode_length_delimited(sign_bytes)?;
+        Ok(true)
+    }
+
+    /// Create signable vector from Proposal.
+    pub fn to_signable_vec(&self, chain_id: ChainId) -> Result<Vec<u8>, DomainTypeError> {
+        CanonicalProposal::new(self.clone(), chain_id).encode_length_delimited_vec()
+    }
+
+    /// Consensus state from this proposal - This doesn't seem to be used anywhere.
```

Comment:
this is used in tmkms via that trait: https://github.com/iqlusioninc/tmkms/blob/develop/src/session.rs#L315

but I guess it could be simpler / clearer without it

Reply:
Yes, I think so too. It's a similar case to the previous. It seems it's generating a `State{}` struct and matching whatever the incoming request was. We can easily do that with a 
`consensus_state = State::from(vote);` or `consensus_state = State::from(proposal);` call. Which tells me that implementing this From trait would be a good idea. I'll check the KMS code a bit further, I'm interested how this state is used.